### PR TITLE
Feature/prod 2359

### PR DESCRIPTION
**Pull request by JasonTLin**
Swagger.json is insecure endpoint

exposed specs endpoint so it can be authorized

@api.specs
@auth.login_required
def custom_swagger_ui():
    """
    Redirect to basic auth, using customer name and token as the password
    """
    return apidoc.specs_for(api)

### CODE REVIEW

File: flask_restplus/api.py [hunk 1]
```diff
@@ -376,6 +381,14 @@ def documentation(self, func):
         self._doc_view = func
         return func
 
+    def specs(self, func):
+        '''A decorator to specify a view function for the specs'''
+        self._hide_specs_url = True
```
Comment (snazzer) [inline flask_restplus/api.py]: Maybe better to default `self._spec_view` to `partial(apidocs.specs_for, self)`. Remove the `_hide_specs_url` so it is always available.

And always inline the swagger data.

I think this results in fewer options, and is closer to original functionality.

File: flask_restplus/api.py [hunk 2]
```diff
@@ -185,6 +189,7 @@ def init_app(self, app, **kwargs):
         self.license = kwargs.get('license', self.license)
         self.license_url = kwargs.get('license_url', self.license_url)
         self._add_specs = kwargs.get('add_specs', True)
+        self._hide_specs_url = kwargs.get('hide_specs_url', False)
```
Comment (snazzer) [inline flask_restplus/api.py]: I don't think you need this flag anymore?

File: flask_restplus/api.py [hunk 3]
```diff
@@ -376,6 +381,15 @@ def documentation(self, func):
         self._doc_view = func
         return func
 
+    def specs(self, func):
+        '''A decorator to specify a view function for the specs'''
+        self._spec_view = partial(apidoc.specs_for, self)
```
Comment (snazzer) [inline flask_restplus/api.py]: This should be in constructor - it's the default.

File: flask_restplus/api.py [hunk 4]
```diff
@@ -376,6 +381,15 @@ def documentation(self, func):
...
+        if func:
```
Comment (snazzer) [inline flask_restplus/api.py]: You can assume func is set.

File: flask_restplus/templates/swagger-ui.html [hunk 5]
```diff
@@ -50,7 +50,11 @@
     <script type="text/javascript">
         window.onload = function() {
             const ui = window.ui = new SwaggerUIBundle({
+                {% if specs_json is defined %}
```
Comment (snazzer) [inline flask_restplus/templates/swagger-ui.html]: So I'm suggesting that you never use the url - just always inline.

File: flask_restplus/api.py [hunk 6]
```diff
@@ -244,7 +249,7 @@ def _register_apidoc(self, app):
         conf['apidoc_registered'] = True
 
     def _register_specs(self, app_or_blueprint):
-        if self._add_specs:
+        if self._add_specs and not self._hide_specs_url:
```
Comment (snazzer) [inline flask_restplus/api.py]: So this can become

```
if self._add_specs:
   self.specs(self._spec_view)
```

or something like that.

**Review by snazzer — CHANGES_REQUESTED** [review]

File: flask_restplus/api.py (hunk 6 above)
Comment (JasonTLin) [reply flask_restplus/api.py]: modified so it just added the url route

**Review by snazzer — APPROVED** [review]
Cool!

**Pull request merged**